torch/profiler/profiler.py

```diff
@@ -0,0 +1,293 @@
+                    p.next_step()
+    """
+    def __init__(
+            self,
+            *,
+            activities: Optional[Iterable[ProfilerActivity]] = None,
+            schedule: Optional[Callable[[int], ProfilerAction]] = None,
+            on_trace_ready: Optional[Callable[..., Any]] = None,
+            record_shapes: bool = False,
+            profile_memory: bool = False,
+            with_stack: bool = False,
+            # deprecated:
+            use_gpu: Optional[bool] = None):
+        if activities:
+            self.activities = activities
+        else:
+            if use_gpu is not None:
+                warn("use_gpu is deprecated, use activities argument instead")
+                self.activities = set([ProfilerActivity.CPU])
+                if use_gpu:
+                    self.activities.add(ProfilerActivity.CUDA)
+            else:
+                raise RuntimeError("Profiler activities are not specified")
+
+        if schedule:
+            self.schedule = schedule
+            # add step markers into the trace and table view
+            self.record_steps = True
+        else:
+            self.schedule = _default_schedule_fn
+            self.record_steps = False
+        self.on_trace_ready = on_trace_ready
+        self.record_shapes = record_shapes
+        self.profile_memory = profile_memory
+        self.with_stack = with_stack
+        self.step_num = 0
+        self.current_action = self.schedule(self.step_num)
+        self.profiler: Optional[prof.profile] = None
+        self.step_rec_fn: Optional[prof.record_function] = None
+
+    def __enter__(self):
+        self._enter_actions()
+        if self.record_steps:
+            self.step_rec_fn = prof.record_function("ProfilerStep#" + str(self.step_num))
+            self.step_rec_fn.__enter__()
+        return self
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        if self.record_steps and self.step_rec_fn:
+            self.step_rec_fn.__exit__(None, None, None)
+        self._exit_actions()
+
+    def next_step(self):
+        """
+        Signals the profiler that the next profiling step has started.
+        """
+        if self.record_steps and self.step_rec_fn:
+            self.step_rec_fn.__exit__(None, None, None)
+        prev_action = self.current_action
+        self.step_num += 1
+        self.current_action = self.schedule(self.step_num)
+
+        if self.current_action == ProfilerAction.NONE:
+            if prev_action == ProfilerAction.NONE:
+                pass
+            elif prev_action == ProfilerAction.WARMUP:
+                warn("Incorrect schedule: WARMUP followed by NONE")
+                self._start_trace()
+                self._stop_trace()
+            elif prev_action == ProfilerAction.RECORD:
+                warn("Incorrect schedule: RECORD followed by NONE")
+                self._stop_trace()
+            else:
+                assert prev_action == ProfilerAction.RECORD_AND_SAVE
```

Comment:
you shouldn't use `assert` for user-induced errors. Change it for `raise` and better error message